pkg/controller/kubemon/statefulset.go

```diff
@@ -0,0 +1,385 @@
+	customProperties := instance.Spec.KubernetesMonitoringSpec.CustomProperties
+	if isCustomPropertiesNilOrEmpty(customProperties) {
+		return nil
+	}
+	return []corev1.Volume{{
+		Name: customproperties.VolumeName,
+		VolumeSource: corev1.VolumeSource{
+			Secret: buildSecretVolumeSource(
+				buildSecretName(instance),
+				customproperties.DataKey,
+				customproperties.DataPath,
+			),
+		},
+	},
+	}
+}
+
+func isCustomPropertiesNilOrEmpty(customProperties *v1alpha1.DynaKubeValueSource) bool {
+	return customProperties == nil ||
+		(customProperties.Value == "" &&
+			customProperties.ValueFrom == "")
+}
+
+func buildSecretVolumeSource(name string, key string, path string) *corev1.SecretVolumeSource {
+	return &corev1.SecretVolumeSource{
+		SecretName: name,
+		Items: []corev1.KeyToPath{
+			{Key: key, Path: path},
+		},
+	}
+}
+
+func buildSecretName(instance *v1alpha1.DynaKube) string {
+	valueFrom := instance.Spec.KubernetesMonitoringSpec.CustomProperties.ValueFrom
+	if valueFrom == "" {
+		valueFrom = fmt.Sprintf("%s-kubernetes-monitoring%s", instance.Name, customproperties.Suffix)
+	}
+	return valueFrom
+}
+
+func buildEnvs(instance *v1alpha1.DynaKube, tenantInfo *dtclient.TenantInfo, kubeSystemUID types.UID) []corev1.EnvVar {
+	return appendProxySettingsEnvVar(instance,
+		append(instance.Spec.KubernetesMonitoringSpec.Env,
+			buildDefaultEnvVars(instance, tenantInfo, kubeSystemUID)...),
+	)
+}
+
+func appendProxySettingsEnvVar(instance *v1alpha1.DynaKube, envVars []corev1.EnvVar) []corev1.EnvVar {
+	proxy := instance.Spec.Proxy
+	if isProxyNilOrEmpty(proxy) {
+		return envVars
+	}
+	return append(envVars, buildProxyEnvVar(proxy))
+}
+
+func buildProxyEnvVar(proxy *v1alpha1.DynaKubeProxy) corev1.EnvVar {
+	if proxy.ValueFrom != "" {
+		return buildProxyEnvVarFromValueSource(proxy)
+	} else {
+		return buildProxyEnvVarFromValue(proxy)
+	}
+}
+
+func buildProxyEnvVarFromValueSource(proxy *v1alpha1.DynaKubeProxy) corev1.EnvVar {
+	return corev1.EnvVar{
+		Name: ProxyEnv,
+		ValueFrom: &corev1.EnvVarSource{
+			SecretKeyRef: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{Name: proxy.ValueFrom},
+				Key:                  ProxyKey,
+			},
+		},
+	}
+}
+
+func buildProxyEnvVarFromValue(proxy *v1alpha1.DynaKubeProxy) corev1.EnvVar {
+	return corev1.EnvVar{
+		Name:  ProxyEnv,
+		Value: proxy.Value,
+	}
+}
+
+func buildDefaultEnvVars(instance *v1alpha1.DynaKube, tenantInfo *dtclient.TenantInfo, kubeSystemUID types.UID) []corev1.EnvVar {
+	return []corev1.EnvVar{
+		{Name: DtTenant, Value: tenantInfo.ID},
+		{Name: DtToken, Value: tenantInfo.Token},
+		{Name: DtServer, Value: tenantInfo.CommunicationEndpoint},
+		{Name: DtCapabilities, Value: CapabilityEnv},
+		{Name: DtIdSeedNamespace, Value: instance.Namespace},
+		{Name: DtIdSeedClusterId, Value: string(kubeSystemUID)},
+	}
+}
+
+func buildArgs(instance *v1alpha1.DynaKube) []string {
```

Comment:
Definitely put them on the same function.

Reply:
merged build*Args functions